src/index.ts

```diff
@@ -90,90 +186,147 @@ export class AssembleSubmitButtonExtension
+  private async submitContentToServer(contents: string) {
+    // post contents to server
+    console.log(contents);
+    const result: ISubmissionResponse = await submit(contents);
+
+    // try to add a message to cell outputs
+    if (result.result) {
+      void showDialog({
+        title: 'Feature submitted successfully',
+        body: new FeatureSubmittedOkayWidget(result.url),
+        buttons: [Dialog.okButton()]
+      });
+    } else {
+      const message =
+        result.message !== undefined && result.message !== null
+          ? `: ${result.message}.`
+          : '.';
+      void showErrorMessage(
+        'Error submitting feature',
+        `Oops - there was a problem submitting your feature${message}`
+      );
+      console.error(result);
+    }
+  }
+
+  private createSliceButton(panel: NotebookPanel) {
+    let button = new ToolbarButton({
+      label: 'Slice',
+      /*icon: new LabIcon({
+        name: 'ballet-icon',
+        svgstr: balletIconSvg
+      }),*/
       onClick: async () => {
-        if (!(await isAuthenticated())) {
-          const popup = window.open(
-            getEndpointUrl('auth/authorize'),
-            '_blank',
-            'width=350,height=600'
-          );
-          // async
-          void request<void>('auth/token', {
-            method: 'POST'
-          });
+        // load current cell
+        let notebook = panel.content;
+        let cells = panel.content.model.cells;
 
-          let authIntervalId;
-          authIntervalId = setInterval(
-            authCallback,
-            0.5 * ONE_SECOND,
-            popup,
-            authIntervalId
+        let content = new Array(cells.length);
+        for (let i = 0; i < cells.length; i++) {
+          content[i] = cells.get(i).value.text;
+        }
+        let ctsJoined = content.join('\n');
+        let ctsSplit = ctsJoined.split('\n');
+
+        // use parse(array) to generate tree
+        let tree = parse(ctsJoined);
+        let loc = this.getLocationFromCurrentCell(
+          notebook.activeCell,
+          ctsSplit
+        );
+
+        let slicedLoc = slice(tree, loc);
+
+        let map = new Map();
+        for (let i = 0; i < slicedLoc.items.length; i++) {
+          const line = slicedLoc.items[i].first_line;
+          // [line - 1] because location type starts at 1, not 0
+          map.set(line, ctsSplit[line - 1]);
+        }
+
+        let arraySorted = [...map.entries()].sort();
+        let finalSlice = [];
+        for (let i = 0; i < arraySorted.length; i++) {
+          finalSlice.push(arraySorted[i][1]);
+        }
+        let result = finalSlice.join('\n');
+
+        const finalDialog = await showDialog({
+          title:
+            'Here is you code slice. Click OK to submit it to your upstream GitHub Repository!',
+          body: new PreviewCodeWidget(result)
+        });
+        if (!finalDialog.button.accept) {
+          return;
+        }
+
+        if (!(await isAuthenticated())) {
```

Comment:
I explicitly put it there since I wanted the user to be able to have look at the slice without necessarily being authenticated. 
Can I leave it like that?

Reply:
yes, that's a fair point. no need for user to authenticate until they confirm they want to submit. leave it like this and I will update later (to consolidate both buttons)